File: python/novie-agent-sdk/src/novie_agent_sdk/call_scope.py
"""Agent 侧拆包 AgentCallScope 的 helper。

Platform 在 invoke payload 里塞 ``inputs["__call_scope__"]``（进 agent
容器时 HttpJsonExternalAgentClient 会把它提成顶层 ``payload["call_scope"]``）。
Agent 侧两条路径都可能看到：

- 进程内（少见）：从 ``inputs["__call_scope__"]`` 拿
- HTTP invoke（常见）：从 ``payload["call_scope"]`` 拿

本 helper 两条都支持。失败时返回 None，agent 走默认行为（shared workspace，
不做 cleanup），不阻塞任务。
"""
from __future__ import annotations

import logging
from typing import Any

from novie_protocol.contracts import AgentCallScope

__all__ = ["extract_call_scope"]

_log = logging.getLogger(__name__)


def extract_call_scope(source: dict[str, Any] | None) -> AgentCallScope | None:
    """从 inputs dict **或** 顶层 payload 里拆出 AgentCallScope。

    容忍两种输入：
    - ``extract_call_scope(inputs)`` — agent 代码拿到的 ``inputs`` 形参
    - ``extract_call_scope(payload)`` — FastAPI / http server 收到的顶层 dict

    返回 None 时 agent 应按保守默认行为跑（即 ``sandbox_isolation=shared`` /
    无 cleanup），而不是 raise。
    """
    if not source:
        return None

    # 优先顶层（http invoke 路径）；其次双下划线（in-process 路径）
    raw = source.get("call_scope")
    if raw is None:
        raw = source.get("__call_scope__")
    if raw is None:
        return None
    if not isinstance(raw, dict):
        _log.warning(
            "call_scope present but not a dict (got %s); ignoring",
            type(raw).__name__,
        )
        return None
    try:
        return AgentCallScope.from_dict(raw)
    except (KeyError, TypeError, ValueError) as exc:
        _log.warning("failed to parse call_scope: %s; ignoring", exc)
        return None

File: rust/novie-agent-sdk/src/call_scope.rs
//! `AgentCallScope` — per-invocation workspace + credential hint contract.
//!
//! Mirrors `novie_protocol.contracts.call_scope`. See the Python module for
//! the full design rationale; in short, the platform does **not** enforce
//! these fields — they are advisory hints the agent uses to pick a workspace
//! layout, cache TTL, and cleanup hook.

use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};

#[derive(Debug, Default, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum WorkspaceScope {
    PerTask,
    PerSession,
    #[default]
    Shared,
    None,
}

#[derive(Debug, Default, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TokenScopeKind {
    PerTask,
    PerSession,
    #[default]
    None,
}

#[derive(Debug, Default, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum CleanupWhen {
    OnStepComplete,
    OnPlanComplete,
    AgentManaged,
    #[default]
    NoCleanup,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct CredentialHint {
    #[serde(default)]
    pub kind: TokenScopeKind,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub ttl_seconds: Option<u64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub authorized_by: Option<String>,
    #[serde(default)]
    pub allowed_resources: Vec<String>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct AgentCallScope {
    #[serde(default)]
    pub workspace_scope: WorkspaceScope,
    #[serde(default)]
    pub cleanup_when: CleanupWhen,
    #[serde(default)]
    pub tenant_id: String,
    #[serde(default)]
    pub workspace_id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub project_id: Option<String>,
    #[serde(default)]
    pub scope_key: String,
    #[serde(default)]
    pub credentials: CredentialHint,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub cleanup_callback_url: Option<String>,
    #[serde(default)]
    pub metadata: Map<String, Value>,
}

/// Pull `AgentCallScope` from either a top-level invoke payload or the
/// `inputs` dict.
///
/// Matches `extract_call_scope(source)` in Python — accepts either path
/// (HTTP invoke vs in-process) and silently returns `None` instead of
/// raising on shape errors so agents fall back to safe defaults.
pub fn extract_call_scope(source: &Value) -> Option<AgentCallScope> {
    let map = source.as_object()?;
    let raw = map
        .get("call_scope")
        .or_else(|| map.get("__call_scope__"))?;
    if !raw.is_object() {
        tracing::warn!(
            type_name = ?raw,
            "call_scope present but not a JSON object; ignoring",
        );
        return None;
    }
    match serde_json::from_value(raw.clone()) {
        Ok(scope) => Some(scope),
        Err(err) => {
            tracing::warn!(error = %err, "failed to parse call_scope; ignoring");
            None
        }
    }
}
